realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -100,18 +100,40 @@ public String getClassName() {
     }
 
     /**
-     * Sets a new name for this RealmObject class. This is equivalent to renaming it.
+     * Sets a new name for this RealmObject class. This is equivalent to renaming it. When {@link RealmObjectSchema#table}
+     * has a primary key, this will transfer the primary key for the new class name.
      *
      * @param className the new name for this class.
+     * @throws IllegalArgumentException if {@param className} is {@code null} or an empty string, or its length exceeds 57 characters.
      * @see RealmSchema#rename(String, String)
      */
     public RealmObjectSchema setClassName(String className) {
         checkEmpty(className);
+        if (57 < className.length()) {
```

Comment:
Actually, this has a deeper issue at Core level. As @beeender points out, Core is supposed to impose 64 chars length of hard-limit on table name including `class_` prefix. The char length is supposed to be checked and thrown when violated due to consistent behavior across bindings.

What happens now is that we can actually insert a class name longer than 57 chars and a table name longer than 64 chars as [_the checker_](https://github.com/realm/realm-core/blob/master/src/realm/replication.cpp#L530) is not working. As soon as the char length checker at Java binding is gone, we can actually do following, which means the bug exists as we speak.

@beeender is to look into this issue in detail. I'd ask we take some time.
<img width="1321" alt="screen shot 2016-07-29 at 6 50 43 pm" src="https://cloud.githubusercontent.com/assets/16395854/17292460/30eb0b96-5825-11e6-8532-1a6942936469.png">